### PR TITLE
Implement NLO charm loop functions for b -> sll

### PR DESCRIPTION
**Pull request by vikusc**
- the difference to the reference values for F29c is large compared to F27c, will have to rearrange terms with the same prefactor
- the large difference in reference values for F29d comes from the region q^2 / m_b^2 ~ 1

### CODE REVIEW

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        complex<double> z(const complex<double> m_c, const complex<double> m_b)
+        {
+            return power_of<2>(m_c) / power_of<2>(m_b);
+        }
+
+        complex<double> xa(const complex<double> m_c, const complex<double> m_b)
+        {
+            if (z(m_c, m_b) == 0.25)
+            {
+                throw InternalError("m_c is not half of m_b");
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: By `is not` you mean `should not`? Is there a problem in this case?

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            }
+            else{
+                return 1.0 / std::sqrt(1 - 4.0 * z(m_c, m_b));#
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: There is a # at the end of the line

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+        complex<double> xd(const complex<double> m_c, const complex<double> m_b)
+        {
+            return xb(m_c, m_b);
+        }
+
+        complex<double> ya(const complex<double> q2, const complex<double> m_c, const complex<double> m_b)
+        {
+            if (1.0 - s(q2, m_b) == 0.0)
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: Why not `s(q2, m_b) == 1.0`?

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            {
+                return 0.0;
+            }
+            else
+            {
+                const complex<double> arg1 = (xconj * a).imag() / denom;
+                const complex<double> arg2 = 1.0 - (xconj * a).imag() / denom;
+                const complex<double> arg3 = - 1.0 - (bconj * a).imag() / denom;
+
+                if ( arg1 < 0.0 || arg2 < 0.0 || arg1 > 0.0)
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: typo? Is it really `arg1` in the end?

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                    throw InternalError("Ill-defined Theta(0)");
+                }
+                else
+                {
+                    return 1.0;
+                }
+            }
+        }
+
+        // Counterterms to the two-loop functions
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: Why do you have this comment?

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 6]
```diff
@@ -9,30 +9,26 @@
...
  * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
- * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
  * details.
  *
  * You should have received a copy of the GNU General Public License along with
- * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
- * Place, Suite 330, Boston, MA  02111-1307  USA
+ * this program; if not, write to the Free Software Foundation, Inc., 59.0 Temple
+ * Place, Suite 330, Boston, MA  02111-1307.0  USA
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc:17]: Can you undo the changes to the warranty statement?

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+#include <eos/utils/log.hh>
+#include <eos/utils/stringify.hh>
+
+#include <cmath>
+#include <complex>
+#include <cstring>
+#include <vector>
+
+#include <gsl/gsl_spline.h>
+#include <gsl/gsl_sf_dilog.h>
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: Can you remove the include statements that are removes in the following commit? Since we will merge all the commits it probably doesn't matter...

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 8]
```diff
@@ -45,102 +41,154 @@ namespace eos
...
 
-        complex<double> xc(const complex<double> m_c, const complex<double> m_b)
+        inline complex<double> x_b(const double & z, const double & feynepsilonhat)
         {
-            return xa(m_c, m_b);
+            return std::sqrt(4.0 * z_eps(z, feynepsilonhat)) - std::sqrt(4.0 * z_eps(z, feynepsilonhat) - 1.0);
         }
 
-        complex<double> xe(const complex<double> m_c, const complex<double> m_b)
+        /* Not needed
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: Can you remove them in this case?

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 9]
```diff
@@ -45,102 +41,154 @@ namespace eos
...
             }
             else
             {
-                return 1.0 / std::sqrt(1.0 - 4.0 * power_of<2>(mc_2) / q2);
+                return 1.0 / std::sqrt(1.0 - 4.0 * z_eps(z, feynepsilonhat) / s_eps(s, feynepsilonhat));
             }
         }
 
-        complex<double> yd(const complex<double> q2, const complex<double> m_c)
+        /* Not needed
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: Ditto

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 10]
```diff
@@ -170,14 +218,33 @@ namespace eos
             }
         }
 
+        // point projected from x in side the triangle 0ab onto the edge ab
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: inside

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 11]
```diff
@@ -170,14 +218,33 @@ namespace eos
...
+        inline complex<double> p(const complex<double> a, const complex<double> b, const complex<double> x)
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: Can you give a better name to this function?

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 12]
```diff
@@ -170,14 +218,33 @@ namespace eos
...
+        {
+            const complex<double> amb = a - b;
+            const complex<double> aconj = std::conj(a);
+            const complex<double> xconj = std::conj(x);
+            const complex<double> aconjb = aconj * b;
+            const double denom = (xconj * amb).imag();
+
+            if (denom == 0.0) // is this the best criterion or should T(a,b,x)!=1.0 be taken?
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: You'll have to answer the question before merging :)

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 13]
```diff
@@ -214,6 +322,254 @@ namespace eos
...
+
+            const complex<double> result = 2.0 / 3.0 + 4.0 * clp.s_eps / clp.z_eps + std::log(power_of<2>(clp.muhat) / clp.z_eps)
+                + (1.0 - 3.0 * power_of<2>(clp.y_c)) / (2.0 * power_of<3>(clp.y_c)) * (std::log(1.0 + clp.y_c) - std::log(1.0 - clp.y_c));
+
+            return 2.0 / 3.0 * q_c * result;
+        }
+
+        complex<double> f190(const CharmLoopsParameters & clp)
+        {
+            const double c_F = 4.0 / 3.0; // SU(3.0) color factor
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc:140]: SU(3.0) !? ^^

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 14]
```diff
@@ -214,6 +322,254 @@ namespace eos
...
+                    + (-144.0 / ye3 + 288.0 / ye - 144.0 * ye) * ln1mxe + (288.0 / ye3 - 576.0 / ye + 288.0 * ye) * lnxe + (-144.0 / ye3 + 288.0 / ye - 144.0 * ye) * ln1pxe
+                    + (-576.0 * ln2 + (192.0 * ln2) / ye4 - (384.0 * ln2) / ye2) * ln1pye)
+                - 576.0 * zeta3 + (192.0 * zeta3) / ye4 - (384.0 * zeta3) / ye2;
+
+            return result / 27.0;
+        }
+
+        complex<double> f19e(const CharmLoopsParameters & clp)
+        {
+            const double c_F = 4.0 / 3.0; // SU(3.0) color factor
```
Comment (mreboud) [inline eos/rare-b-decays/charm-loops-analytic.cc]: Ditto

File: eos/rare-b-decays/Makefile.am [hunk 15]
```diff
@@ -29,6 +29,14 @@ libeosrarebdecays_la_SOURCES = \
 	bs-to-phi-ll-bfs2004.cc bs-to-phi-ll-bfs2004.hh \
 	bs-to-phi-ll-gvdv2020.cc bs-to-phi-ll-gvdv2020.hh \
 	charm-loops.cc charm-loops.hh \
+	charm-loops-analytic.cc charm-loops-impl.hh \
+    charm-loops-analytic-F27b.cc \
```
Comment (mreboud) [inline eos/rare-b-decays/Makefile.am:33]: Indentation are wrong on github, you probably have a tabulation somewhere

**Review by mreboud — COMMENTED** [review]
I started reviewing, but the modification tree is not linear so I left a lot of comments that are made obsolete by newer commits.

File: eos/rare-b-decays/charm-loops-analytic.cc [hunk 16]
```diff
@@ -9,30 +9,26 @@
...
  */
 
 #include <eos/maths/power-of.hh>
 #include <eos/maths/polylog.hh>
-#include <eos/maths/multiplepolylog_li22.hh>
+#include <eos/maths/multiplepolylog-li22.hh>
+
 #include <eos/rare-b-decays/charm-loops.hh>
-#include <eos/rare-b-decays/long-distance.hh>
+
```
[the remark this answers is not in the document]
Comment (vikusc) [reply eos/rare-b-decays/charm-loops-analytic.cc:25]: Is there still an unnecessary line?

File: eos/rare-b-decays/charm-loops-analytic.cc (hunk 5 above)
Comment (vikusc) [reply eos/rare-b-decays/charm-loops-analytic.cc]: It's meant to structure the file better

**Comment by mreboud** [comment]
Can we merge this @dvandyk @vikusc ?

**Comment by dvandyk** [comment]
Yes please!

**Review by dvandyk — APPROVED** [review]
